plugins/touchegg/Client.vala

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2020 elementary, Inc (https://elementary.io)
+ *           2020 José Expósito <jose.exposito89@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace Gala.Plugins.Touchegg {
+    /**
+     * Daemon event type.
+     */
+     private enum GestureEventType {
+        UNKNOWN = 0,
+        BEGIN = 1,
+        UPDATE = 2,
+        END = 3,
+    }
+
+    /**
+     * Daemon event.
+     */
+    private struct GestureEvent {
+        public uint32 event_size;
+        public GestureEventType event_type;
+        public GestureType type;
+        public GestureDirection direction;
+        public int percentage;
+        public int fingers;
+        public uint64 elapsed_time;
+        public DeviceType performed_on_device_type;
+    }
+
+    /**
+     * This class connects to the Touchégg daemon to receive touch events.
+     * See: https://github.com/JoseExposito/touchegg
+     */
+    public class Client {
+        public signal void on_gesture_begin (Gesture gesture);
+        public signal void on_gesture_update (Gesture gesture);
+        public signal void on_gesture_end (Gesture gesture);
+
+        /**
+         * Maximum number of reconnection attemps to the daemon.
+         */
+        private const int MAX_RECONNECTION_ATTEMPS = 5;
+
+        /**
+         * Socket used to connect to the daemon.
+         */
+        private Socket socket = null;
+
+        /**
+         * Current number of reconnection attemps.
+         */
+        private int reconnection_attemps = 0;
+
+        /**
+         * Stuct to store the received event. It is usefull to keep it to be able to finish ongoing
+         * actions in case of disconnection
+         */
+        private GestureEvent *event = null;
+
+        /**
+         * Start receiving gestures.
+         */
+        public void run () throws IOError {
+            new Thread<void*> (null, this.recive_events);
+        }
+
+        private void* recive_events () {
+            uint8[] event_buffer = new uint8[sizeof (GestureEvent)];
```

Comment:
I don't think we need to allocate this buffer using `new`, I think there are better ways to do this (not sure which will compile though):
1. `uint8[sizeof(GestureEvent)] event_buffer;`
2. `socket.receive ((uint8[])event);`
3. If none of this works, I'm pretty sure you can allocate this buffer in the constructor and reuse it here, simply zeroing it before calling receive.

Reply:
The buffer only gets allocated once because on the next line the client starts its event loop.
Is Vala really slow allocating memory on the heap? This is the first time I write Vala so I may not know the advantage of using memory in the stack vs the heap.